<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Scientific Computation at large - WGS</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Scientific Computation at large</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html">About</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">WGS</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Omics Analysis</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_0.html" class="sidebar-item-text sidebar-link">Intro</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_pyro_cov.html" class="sidebar-item-text sidebar-link">pyro_cov</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_varscan.html" class="sidebar-item-text sidebar-link">VarScan</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_wgs.html" class="sidebar-item-text sidebar-link active">WGS</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">Protein Design</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_0.html" class="sidebar-item-text sidebar-link">Intro</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_AlphaFold.html" class="sidebar-item-text sidebar-link">AlphaFold</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_AFdesign.html" class="sidebar-item-text sidebar-link">AFdesign</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_binding_ddg.html" class="sidebar-item-text sidebar-link">binding_ddG</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_chroma.html" class="sidebar-item-text sidebar-link">chroma</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_MPNN.html" class="sidebar-item-text sidebar-link">ProteinMPNN</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_RFdiffusion.html" class="sidebar-item-text sidebar-link">RFdiffusion</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">Micromoecule Design</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_0.html" class="sidebar-item-text sidebar-link">Intro</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_difflinker.html" class="sidebar-item-text sidebar-link">DiffLinker</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">Molecular Dynamics</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_0.html" class="sidebar-item-text sidebar-link">Molecular dynamics</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">Machine Learning</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_0.html" class="sidebar-item-text sidebar-link">Machine learning</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">New Technologies</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_0.html" class="sidebar-item-text sidebar-link">New technologies</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tools" id="toc-tools" class="nav-link active" data-scroll-target="#tools">Tools</a></li>
  <li><a href="#reference-genome" id="toc-reference-genome" class="nav-link" data-scroll-target="#reference-genome">Reference genome</a></li>
  <li><a href="#reads" id="toc-reads" class="nav-link" data-scroll-target="#reads">Reads</a></li>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">Quality control</a></li>
  <li><a href="#align-reads-to-reference-genome" id="toc-align-reads-to-reference-genome" class="nav-link" data-scroll-target="#align-reads-to-reference-genome">Align reads to reference genome</a>
  <ul class="collapse">
  <li><a href="#view-sam-file" id="toc-view-sam-file" class="nav-link" data-scroll-target="#view-sam-file">View SAM file</a></li>
  </ul></li>
  <li><a href="#convert-sam-to-bam" id="toc-convert-sam-to-bam" class="nav-link" data-scroll-target="#convert-sam-to-bam">Convert SAM to BAM</a></li>
  <li><a href="#sort-bam" id="toc-sort-bam" class="nav-link" data-scroll-target="#sort-bam">Sort BAM</a></li>
  <li><a href="#mark-duplicates" id="toc-mark-duplicates" class="nav-link" data-scroll-target="#mark-duplicates">Mark duplicates</a></li>
  <li><a href="#index-bam" id="toc-index-bam" class="nav-link" data-scroll-target="#index-bam">Index BAM</a></li>
  <li><a href="#base-quality-score-recalibration" id="toc-base-quality-score-recalibration" class="nav-link" data-scroll-target="#base-quality-score-recalibration">Base quality score recalibration</a></li>
  <li><a href="#haplotype-caller" id="toc-haplotype-caller" class="nav-link" data-scroll-target="#haplotype-caller">Haplotype caller</a></li>
  <li><a href="#variant-filtering" id="toc-variant-filtering" class="nav-link" data-scroll-target="#variant-filtering">Variant filtering</a></li>
  <li><a href="#variant-annotation" id="toc-variant-annotation" class="nav-link" data-scroll-target="#variant-annotation">Variant annotation</a></li>
  <li><a href="#variant-recalibration" id="toc-variant-recalibration" class="nav-link" data-scroll-target="#variant-recalibration">Variant recalibration</a></li>
  <li><a href="#apply-variant-recalibration" id="toc-apply-variant-recalibration" class="nav-link" data-scroll-target="#apply-variant-recalibration">Apply variant recalibration</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">WGS</h1>
<p class="subtitle lead">WGS variant calling using GATK</p>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<ul class="task-list">
<li><input type="checkbox" disabled="">clarify the choice of reference genome</li>
<li><input type="checkbox" disabled="">clarify the choice of known sites</li>
<li><input type="checkbox" disabled="">modify the code to suit for multiple samples</li>
</ul>
<section id="tools" class="level2">
<h2 class="anchored" data-anchor-id="tools">Tools</h2>
<p>Tools necessary for WGS variant calling using GATK include:</p>
<p>Picard is a Java-based toolkit for manipulating high-throughput sequencing (HTS) data and formats such as SAM/BAM/CRAM and VCF.</p>
<p>GATK is a toolkit for variant discovery in high-throughput sequencing data. It is designed to work best with data generated by next-generation DNA sequencing technologies, but can be applied to other types of data as well.</p>
<p>Samtools is a suite of programs for interacting with high-throughput sequencing data. It provides various utilities for post-processing alignments in the SAM, BAM and CRAM formats, such as indexing, variant calling, viewing alignments, and generating alignments in a per-position format. It also includes a large number of utilities for manipulating alignments in a variety of different ways, for example, to extract all reads that map to a particular region of the genome.</p>
<p>BWA is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome. It consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM. The first algorithm is designed for Illumina sequence reads up to 100bp, while the rest two for longer sequences ranged from 70bp to 1Mbp. BWA-MEM and BWA-SW share similar features such as long-read support and split alignment, but BWA-MEM, which is the latest, is generally recommended for high-quality queries as it is faster and more accurate. BWA-MEM also has better performance than BWA-backtrack for 70-100bp Illumina reads. For all the algorithms, BWA first needs to construct the FM-index for the reference genome (the index command). Alignment algorithms are invoked with different sub-commands: aln/samse/sampe for BWA-backtrack, bwasw for BWA-SW and mem for the BWA-MEM algorithm.</p>
<p>FastQC is a quality control tool for high throughput sequence data. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis. It produces a simple HTML file with a summary of the results of the analysis, which can be viewed in a web browser.</p>
<p>Other tools that are useful for WGS variant calling include: - Freebayes, a Bayesian genetic variant detector designed to find small polymorphisms, specifically SNPs (single-nucleotide polymorphisms), indels (insertions and deletions), MNPs (multi-nucleotide polymorphisms), and complex events (composite insertion and substitution events) - bcftools, a set of utilities that manipulate variant calls in the Variant Call Format (VCF) and its binary counterpart BCF - vcftools, a set of tools written in Perl and C++ for working with VCF files, such as those generated by the 1000 Genomes Project - bedtools, a suite of utilities for genome arithmetic - tabix, a generic indexer for TAB-delimited genome position files - bgzip, a small utility that enables tabix to work with gzip-compressed files</p>
</section>
<section id="reference-genome" class="level2">
<h2 class="anchored" data-anchor-id="reference-genome">Reference genome</h2>
<p>define the reference genome directory and files</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>REF_DIR <span class="op">=</span> <span class="st">"/mnt/nas/wgs/hg38"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>REF_URL <span class="op">=</span> <span class="st">"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>REF_FILE <span class="op">=</span> os.path.join(REF_DIR, <span class="st">"hg38.fa"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>KNOWN_SITES <span class="op">=</span> os.path.join(REF_DIR, <span class="st">"Homo_sapiens_assembly38.dbsnp138.vcf"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>KNOWN_SITES_IDX <span class="op">=</span> os.path.join(REF_DIR, <span class="st">"Homos_sapiens_assembly38.dbsnp138.vcf.idx"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>download reference genome and index it</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the reference genome if the file doesn't exist</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(REF_DIR, <span class="st">"hg38.fa"</span>)):</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget <span class="op">-</span>O $REF_DIR<span class="op">/</span>hg38.fa.gz $REF_URL</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>gunzip $REF_DIR<span class="op">/</span>hg38.fa.gz</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># index the reference genome file before running haplotype caller</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(REF_DIR, <span class="st">"hg38.fa.fai"</span>)):</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>samtools faidx $REF_DIR<span class="op">/</span>hg38.fa</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dictionary file for the reference genome if it doesn't exist</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(REF_DIR, <span class="st">"hg38.dict"</span>)):</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>gatk CreateSequenceDictionary R<span class="op">=</span>$REF_DIR<span class="op">/</span>hg38.fa O<span class="op">=</span>$REF_DIR<span class="op">/</span>hg38.<span class="bu">dict</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="co"># index the reference genome file using BWA, if the index files don't exist</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(REF_DIR, <span class="st">"hg38.fa.sa"</span>)):</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>bwa index $REF_FILE</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>SITE_URL <span class="op">=</span> <span class="st">"https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf"</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>SITE_IDX_URL <span class="op">=</span> <span class="st">"https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.idx"</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Download known sites files for BQSR (Base Quality Score Recalibration) from GATK resource bundle if they don't exist</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(REF_DIR, <span class="st">"Homo_sapiens_assembly38.dbsnp138.vcf"</span>)):</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget <span class="op">-</span>O $REF_DIR $SITE_URL</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(REF_DIR, <span class="st">"Homo_sapiens_assembly38.dbsnp138.vcf.idx"</span>)):</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget <span class="op">-</span>O $REF_DIR $SITE_IDX_URL</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="co"># show the contents of the reference directory</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ls <span class="op">-</span>halt $REF_DIR</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="reads" class="level2">
<h2 class="anchored" data-anchor-id="reads">Reads</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>READS_DIR <span class="op">=</span> <span class="st">"/mnt/nas/wgs/RDD2206P0001"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>READS_FILE1 <span class="op">=</span> os.path.join(READS_DIR, <span class="st">"RDD2206P0001-WGS-1_R1.fastq.gz"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>READS_FILE2 <span class="op">=</span> os.path.join(READS_DIR, <span class="st">"RDD2206P0001-WGS-1_R2.fastq.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-control" class="level2">
<h2 class="anchored" data-anchor-id="quality-control">Quality control</h2>
<p>This is a quick overview of the quality of the data. We will use FastQC to generate a report for each sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FastQC on the reads</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>fastqc $READS_FILE1 $READS_FILE2 <span class="op">-</span>o $READS_DIR</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="align-reads-to-reference-genome" class="level2">
<h2 class="anchored" data-anchor-id="align-reads-to-reference-genome">Align reads to reference genome</h2>
<p>the aligner <code>bwa aln</code> can trim the reads to remove low quality bases at the end of the reads. This is done using the <code>-q</code> option. The default value is 15.</p>
<p>The BWA-MEM algorithm performs local alignment. It may produce multiple primary alignments for different part of a query sequence. This is a crucial feature for long sequences. However, some tools such as Picard’s <code>markDuplicates</code> does not work with split alignments. One may consider to use option <code>-M</code> to flag shorter split hits as secondary. This option is not recommended for Illumina short reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Align the reads to the reference genome</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bwa mem <span class="op">-</span>t <span class="dv">50</span> <span class="op">-</span>M <span class="op">-</span>R <span class="st">"@RG</span><span class="ch">\t</span><span class="st">ID:RDD2206P0001</span><span class="ch">\t</span><span class="st">SM:RDD2206P0001"</span> $REF_FILE $READS_FILE1 $READS_FILE2 <span class="op">&gt;</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sam</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>-R</code> marks the complete read group header line. ’ can be used in STR and will be converted to a TAB in the output SAM. The read group ID will be attached to every read in the output. The read group ID is used to identify the read group in the output file. The read group sample name is used to identify the sample in the output file.</p>
<section id="view-sam-file" class="level3">
<h3 class="anchored" data-anchor-id="view-sam-file">View SAM file</h3>
<p>the SAM file is a text file that contains the alignment of the reads to the reference genome. SAM stands for Sequence Alignment/Map format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the first 10 lines of the SAM file</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>head <span class="op">-</span>n <span class="dv">10</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sam</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can view the flag statistics of the SAM file using <code>samtools flagstat</code> command. The output of this command is a table that shows the number of reads that have a particular flag.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the flag statistics</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>samtools flagstat $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sam</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The SAM file contains the following flags:</p>
<ul>
<li>0x1: template having multiple segments in sequencing</li>
<li>0x2: each segment properly aligned according to the aligner</li>
<li>0x4: segment unmapped</li>
<li>0x8: next segment in the template unmapped</li>
<li>0x10: SEQ being reverse complemented</li>
<li>0x20: SEQ of the next segment in the template being reversed</li>
<li>0x40: the first segment in the template</li>
<li>0x80: the last segment in the template</li>
<li>0x100: secondary alignment</li>
<li>0x200: not passing quality controls</li>
<li>0x400: PCR or optical duplicate</li>
<li>0x800: supplementary alignment</li>
</ul>
<p>The SAM file contains the following fields:</p>
<ul>
<li>QNAME: Query template NAME</li>
<li>FLAG: bitwise FLAG</li>
<li>RNAME: Reference sequence NAME</li>
<li>POS: 1-based leftmost mapping POSition</li>
<li>MAPQ: MAPping Quality</li>
<li>CIGAR: CIGAR string</li>
<li>RNEXT: Ref. name of the mate/next read</li>
<li>PNEXT: Position of the mate/next read</li>
<li>TLEN: observed Template LENgth</li>
<li>SEQ: segment SEQuence</li>
<li>QUAL: ASCII of Phred-scaled base QUALity+33</li>
</ul>
<p>The SAM file needs to be sorted by the read name. This is done to make the file compatible with <code>MarkDuplicatesSpark</code> and <code>HaplotypeCallerSpark</code>.</p>
</section>
</section>
<section id="convert-sam-to-bam" class="level2">
<h2 class="anchored" data-anchor-id="convert-sam-to-bam">Convert SAM to BAM</h2>
<p>the BAM file is a binary file that contains the alignment of the reads to the reference genome. BAM stands for Binary Alignment/Map format. BAM is a compressed binary version of SAM.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>samtools view <span class="op">-</span>bS $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sam</span> <span class="op">&gt;</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">bam</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sort-bam" class="level2">
<h2 class="anchored" data-anchor-id="sort-bam">Sort BAM</h2>
<p>The BAM file is sorted by the read name. This is done to make the file compatible with <code>MarkDuplicatesSpark</code> and <code>HaplotypeCallerSpark</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>samtools sort <span class="op">-@</span> <span class="dv">50</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">bam</span> <span class="op">-</span>o $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the flag statistics</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>samtools flagstat $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="mark-duplicates" class="level2">
<h2 class="anchored" data-anchor-id="mark-duplicates">Mark duplicates</h2>
<p>Mark duplicates in the BAM file. This is done to remove PCR duplicates. PCR duplicates are reads that are generated from the same DNA fragment. PCR duplicates are generated during the PCR amplification step. PCR duplicates are removed to avoid overestimation of the coverage of the genome.</p>
<p><code>MarkDuplicatesSpark</code> is a Spark version of <code>MarkDuplicates</code>. It is faster than <code>MarkDuplicates</code> and can be run on a cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk MarkDuplicatesSpark <span class="op">-</span>I $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.bam <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.bam <span class="op">-</span>M $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.metrics</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the flag statistics</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>samtools flagstat $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="index-bam" class="level2">
<h2 class="anchored" data-anchor-id="index-bam">Index BAM</h2>
<p>Index the BAM file. This is done to make the file compatible with <code>HaplotypeCallerSpark</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>samtools index $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="base-quality-score-recalibration" class="level2">
<h2 class="anchored" data-anchor-id="base-quality-score-recalibration">Base quality score recalibration</h2>
<p>Base quality score recalibration (BQSR) is a process that recalibrates the base quality scores of the reads. BQSR is done to correct for systematic errors in the base quality scores. BQSR is done using the <code>BaseRecalibrator</code> and <code>ApplyBQSR</code> tools. The <code>BaseRecalibrator</code> tool generates a recalibration table. The <code>ApplyBQSR</code> tool applies the recalibration table to the BAM file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk BaseRecalibrator <span class="op">-</span>I $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.bam <span class="op">-</span>R $REF_FILE <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal_data.table <span class="op">--</span>known<span class="op">-</span>sites $KNOWN_SITES</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk ApplyBQSR <span class="op">-</span>R $REF_FILE <span class="op">-</span>I $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.bam <span class="op">--</span>bqsr<span class="op">-</span>recal<span class="op">-</span><span class="bu">file</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal_data.table <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="haplotype-caller" class="level2">
<h2 class="anchored" data-anchor-id="haplotype-caller">Haplotype caller</h2>
<p>The HaplotypeCaller tool is used to call variants. The HaplotypeCaller tool is a Spark version of the HaplotypeCaller tool. It is faster than the HaplotypeCaller tool and can be run on a cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk HaplotypeCaller <span class="op">-</span>R $REF_FILE <span class="op">-</span>I $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.bam <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="variant-filtering" class="level2">
<h2 class="anchored" data-anchor-id="variant-filtering">Variant filtering</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk VariantFiltration <span class="op">-</span>R $REF_FILE <span class="op">-</span>V $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.vcf <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.vcf <span class="op">--</span><span class="bu">filter</span><span class="op">-</span>expression <span class="st">"QD &lt; 2.0"</span> <span class="op">--</span><span class="bu">filter</span><span class="op">-</span>name <span class="st">"QD2"</span> <span class="op">--</span><span class="bu">filter</span><span class="op">-</span>expression <span class="st">"FS &gt; 60.0"</span> <span class="op">--</span><span class="bu">filter</span><span class="op">-</span>name <span class="st">"FS60"</span> <span class="op">--</span><span class="bu">filter</span><span class="op">-</span>expression <span class="st">"MQ &lt; 40</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="variant-annotation" class="level2">
<h2 class="anchored" data-anchor-id="variant-annotation">Variant annotation</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk VariantAnnotator <span class="op">-</span>R $REF_FILE <span class="op">-</span>V $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.vcf <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.vcf <span class="op">--</span>dbsnp $KNOWN_SITES</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="variant-recalibration" class="level2">
<h2 class="anchored" data-anchor-id="variant-recalibration">Variant recalibration</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk VariantRecalibrator <span class="op">-</span>R $REF_FILE <span class="op">-</span>V $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.vcf <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.recal.vcf <span class="op">--</span>tranches<span class="op">-</span><span class="bu">file</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.recal.tranches <span class="op">--</span>rscript<span class="op">-</span><span class="bu">file</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.recal.plots.R <span class="op">--</span>resource:hapmap,known<span class="op">=</span>false,training<span class="op">=</span>true,truth<span class="op">=</span>true,prior<span class="op">=</span><span class="fl">15.0</span> $HAPMAP <span class="op">--</span>resource:omni,known<span class="op">=</span>false,training<span class="op">=</span>true,truth<span class="op">=</span>false,prior<span class="op">=</span><span class="fl">12.0</span> $OMNI <span class="op">--</span>resource:<span class="dv">1000</span><span class="er">G</span>,known<span class="op">=</span>false,training<span class="op">=</span>true,truth<span class="op">=</span>false,prior<span class="op">=</span><span class="fl">10.0</span> $KG <span class="op">--</span>resource:dbsnp,known<span class="op">=</span>true,training<span class="op">=</span>false,truth<span class="op">=</span>false,prior<span class="op">=</span><span class="fl">2.0</span> $KNOWN_SITES <span class="op">--</span>an QD <span class="op">--</span>an MQ <span class="op">--</span>an MQRankSum <span class="op">--</span>an ReadPosRankSum <span class="op">--</span>an FS <span class="op">--</span>an SOR <span class="op">--</span>mode SNP <span class="op">--</span>tranche <span class="fl">100.0</span> <span class="op">--</span>tranche <span class="fl">99.9</span> <span class="op">--</span>tranche <span class="fl">99.0</span> <span class="op">--</span>tranche <span class="fl">90.0</span> <span class="op">--</span><span class="bu">max</span><span class="op">-</span>gaussians <span class="dv">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="apply-variant-recalibration" class="level2">
<h2 class="anchored" data-anchor-id="apply-variant-recalibration">Apply variant recalibration</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gatk ApplyVQSR <span class="op">-</span>R $REF_FILE <span class="op">-</span>V $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.vcf <span class="op">-</span>O $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.recal.vcf <span class="op">--</span>ts<span class="op">-</span><span class="bu">filter</span><span class="op">-</span>level <span class="fl">99.0</span> <span class="op">--</span>tranches<span class="op">-</span><span class="bu">file</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.recal.tranches <span class="op">--</span>recal<span class="op">-</span><span class="bu">file</span> $READS_DIR<span class="op">/</span>RDD2206P0001<span class="op">-</span>WGS<span class="op">-</span><span class="fl">1.</span><span class="er">sorted</span>.dedup.recal.filtered.annotated.recal.vcf <span class="op">--</span>mode SNP</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>